libraries/custom_appbase/include/eosio/chain/exec_pri_queue.hpp

```diff
@@ -0,0 +1,233 @@
+#pragma once
+#include <boost/asio.hpp>
+
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+
+namespace appbase {
+// adapted from: https://www.boost.org/doc/libs/1_69_0/doc/html/boost_asio/example/cpp11/invocation/prioritised_handlers.cpp
+
+// Locking has to be coordinated by caller, use with care.
+class exec_pri_queue : public boost::asio::execution_context
+{
+public:
+
+   void enable_locking(uint32_t num_threads, std::function<bool()> should_exit) {
+      assert(num_threads > 0 && num_waiting_ == 0);
+      lock_enabled_ = true;
+      max_waiting_ = num_threads;
+      should_exit_ = std::move(should_exit);
+      exiting_blocking_ = false;
+   }
+
+   void disable_locking() {
+      lock_enabled_ = false;
+      should_exit_ = [](){ assert(false); return true; }; // should not be called when locking is disabled
```

Comment:
Sure, that makes sense. This should never happen, but I guess it is better to have the abort disabled in release mode.

Reply:
A slightly more efficient approach is we just remove the line. `should_exit_` is not called and if it was it would most likely return true because of deadline being exceeded.

Reply:
Actually I prefer it as is, because if the `assert()` is triggered it will alert us that something is not working as expected.